File: src/Lecture3/resizeCircularArray/Main.java
package resize_circular_array;

import java.util.Arrays;

public class Main {
  public static void main(String[] args) {
    int[] arr = {3, 4, 5, 1, 2};
    var resize = new Resize(arr);
    arr = resize.expand(3, 5, 2);
    System.out.println(Arrays.toString(arr));
  }
}

class Resize {
  int[] nativeArr;

  public Resize(int[] arr) {
    this.nativeArr = arr;
  }

  public int[] expand(int start, int size, int increaseBy) {
    int[] resizedArr = new int[lenOf(nativeArr) + increaseBy];
    var resizeIndex = start;
    for (var i = 0; i < size; i++) {
      resizedArr[resizeIndex] = nativeArr[start];
      start = (start + 1) % lenOf(nativeArr);
      resizeIndex = (resizeIndex + 1) % lenOf(resizedArr);
    }
    return resizedArr;
  }

  private int lenOf(int[] arr) {
    return arr.length;
  }
}

File: src/Lecture3/resizeCircularArray/main.py
import array as arr

class Resize():
  def __init__(self, circ):
    self.circ = circ

  def expand(self, increaseBy, start):
    brr = arr.array('i', (len(self.circ) + increaseBy) * [0])
    newIndex = natIndex = start
    for i in range(len(self.circ)):
      brr[newIndex] = self.circ[natIndex]
      natIndex = (natIndex + 1) % len(self.circ)
      newIndex = (newIndex + 1) % len(brr)
    return brr


if __name__ == "__main__":
  resize = Resize(arr.array('i',[3, 4, 5, 1, 2]))
  circ = resize.expand(2, 3)
  print(circ)
